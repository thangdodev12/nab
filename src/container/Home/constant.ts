export const TYPES = {
  LOAD_DATA_START: "LOAD_DATA_START",
  LOAD_DATA_COMPLETE: "LOAD_DATA_COMPLETE",
  LOAD_DATA_FAILED: "LOAD_DATA_FAILED",
  GET_LOCATION_START: "GET_LOCATION_START",
  GET_LOCATION_COMPLETE: "GET_LOCATION_COMPLETE",
};