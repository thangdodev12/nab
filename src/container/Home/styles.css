.app {
  padding: 2em 1em;
}
.search__input {
  font-size: 0.75em;
  border: 1px solid #ccc;
  padding: 0.5em 1em;
  border-radius: 4px;
  width: 100%;
  max-width: 320px;
  margin-right: 1em;
}
.search__button {
  padding: 0.5em 1em;
  font-size: 0.75em;
  background-color: #337ab7;
  color: #fff;
  border: none;
  border-radius: 4px;
}
.search__button:hover {
  opacity: 0.75;
}
.content__body {
  display: flex;
  flex-direction: column;
}
.content__header {
  border-bottom: 1px solid #337ab7;
}

.content__header__title {
  font-size: 1.5em;
  color: #337ab7;
}

.content__header__date {
  font-size: 0.75em;
}

@media (min-width: 768px) {
  .content__body {
    flex-direction: row;
    flex-wrap: wrap;
  }
}