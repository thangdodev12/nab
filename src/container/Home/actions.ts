import { TYPES } from "./constant";

export const loadData = (id?: string) => {
  return {
    type: TYPES.LOAD_DATA_START,
    payload: id,
  };
};

export const updateData = (payload: any) => {
  return {
    type: TYPES.LOAD_DATA_COMPLETE,
    payload,
  };
};

export const updateDataFailed = () => {
  return {
    type: TYPES.LOAD_DATA_FAILED,
  };
};

export const getLocationWeather = (payload: string) => {
  return {
    type: TYPES.GET_LOCATION_START,
    payload,
  };
}