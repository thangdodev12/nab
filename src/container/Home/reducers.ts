import { AnyAction } from "redux";
import { TYPES } from "./constant";

interface IAppState {
  data: any
  locations: any,
  loading: boolean
  error: boolean
}

export const initialState = {
  loading: false,
  data: [],
  locations: [],
  error: false,
};

const reducers = (state: IAppState = initialState,  actions: AnyAction) => {
  switch (actions.type) {
    case TYPES.LOAD_DATA_START:
      return { ...state, loading: true, error: false };
    case TYPES.LOAD_DATA_COMPLETE:
      return { ...state, data: actions.payload, loading: false };
    case TYPES.LOAD_DATA_FAILED:
      return { ...state, error: true, loading: false };
    case TYPES.GET_LOCATION_START:
      return { ...state, loading: true, error: false };
    case TYPES.GET_LOCATION_COMPLETE:
      return { ...state, locations: actions.payload, loading: false };
    default: {
      return state;
    }
  }
}

export default reducers;