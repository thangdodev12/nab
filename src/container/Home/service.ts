import axios from "axios";
import { API } from '../../constant';

export  const getWeatherData = async (id?: string) => {
  const { data } = await axios.get(`${API}/weather/${id || 1252431}`);
  return data;
}

export  const getLocationService = async (value: string) => {
  const { data } = await axios.get(`${API}/location?query=${value}`);
  return data;
} 