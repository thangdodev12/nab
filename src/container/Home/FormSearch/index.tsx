import { useCallback } from 'react';
import { connect } from 'react-redux';
import { getLocationWeather as getLocationWeatherAction } from '../actions';
import './styles.css';

const FormSearch = ({ loading, getLocationWeather }: any) => {
  const onSubmit = useCallback((e) => {
    e.preventDefault();
    getLocationWeather(e.target.search.value);
  }, [getLocationWeather]);
  return (
    <form className="search" onSubmit={onSubmit}>
      <input type="search" className="search__input" name="search" placeholder="ex: Ho Chi Minh" />
      <button type="submit" className="search__button" disabled={loading}>Search</button>
    </form>
  );
}

const mapDispatchToProps = (dispatch: any) => ({
  getLocationWeather: (query: string) => dispatch(getLocationWeatherAction(query)),
});

export default connect(null, mapDispatchToProps)(FormSearch);
