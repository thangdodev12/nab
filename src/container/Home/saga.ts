import { SagaIterator } from "redux-saga";
import { call, put, takeLatest } from "redux-saga/effects";
import { updateData, loadData, updateDataFailed } from "./actions";
import { TYPES } from "./constant";
import { getWeatherData, getLocationService } from './service';

export function* loadDataStartSaga(action: any): SagaIterator {
  try {
    const { data } = yield call(getWeatherData, action.payload);
    yield put(updateData(data));
  } catch (error) {
    yield put(updateDataFailed());
  }
};

export function* getLocationStartSaga(action: any): SagaIterator {
  try {
    const { data } = yield call(getLocationService, action.payload);
    const location = data[0];
    yield put(loadData(location.woeid));
  } catch (error) {
    yield put(updateDataFailed());
  }
  
};

export function* initHomeSaga() {
  yield takeLatest(TYPES.LOAD_DATA_START, loadDataStartSaga);
  yield takeLatest(TYPES.GET_LOCATION_START, getLocationStartSaga);
}