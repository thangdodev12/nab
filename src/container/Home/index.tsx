/* eslint-disable react-hooks/exhaustive-deps */
import { FC } from 'react';
import moment from 'moment';
import { useEffect } from 'react';
import { connect, useDispatch } from 'react-redux';
import { IWeather } from '../../constant/interface';
import Weather from '../../component/Weather';
import Loading from '../../component/Loading';
import { loadData } from './actions';
import FormSearch from './FormSearch';
import './styles.css';

interface IProps {
  loading?: boolean,
  data?: IWeather,
}

export const Home: FC<IProps> = ({ loading, data }: IProps) => {
  const { consolidatedWeather = [], title, time } = data as IWeather
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(loadData());
  }, []);
  const [currentWeahter, ...weather] = consolidatedWeather;
  return (
    <div className="app">
      <FormSearch loading={loading} />
        {loading ? (
          <Loading />
        ) : (
          <div className="content">
            <div className="content__header">
              <h1 className="content__header__title">{title}</h1>
              <p className="content__header__date"><strong>Date: </strong>{moment(time).format('DD/MM/YYYY')}</p>
            </div>
            <Weather info={{...currentWeahter, title: 'Today'}} />
            <div className="content__body">
              {weather.map((info: any) => (
                <Weather info={info} key={info.id} />
              ))}
            </div>
          </div>
        )}
    </div>
  );
}

const mapStateToProps = ({ home }: any) => {
  return {
    loading: home.loading,
    data: home.data,
  };
};

export default connect(mapStateToProps, null)(Home);
