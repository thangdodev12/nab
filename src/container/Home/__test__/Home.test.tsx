import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Home from '..';

const mockStore = configureStore([]);

describe('Home', () => {
  let props, store;
  beforeEach(() => {
    store = mockStore({
      home: {
        data: {
          title: 'Ho Chi Minh',
          consolidatedWeather: [],
          time: '2021-01-01'
        }
      },
    });
    props = {
      loading: false,
      data: {
        title: 'Ho Chi Minh',
        consolidatedWeather: [],
        time: '2021-01-01'
      }
    }
  });

  it('should render match  snapshot', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Home {...props} />
      </Provider>,
    );
    const snapshot = component.toJSON();
    expect(snapshot).toMatchSnapshot();
  })
})
