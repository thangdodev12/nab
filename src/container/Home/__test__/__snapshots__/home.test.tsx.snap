// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render match  snapshot 1`] = `
<div
  className="app"
>
  <form
    className="search"
    onSubmit={[Function]}
  >
    <input
      className="search__input"
      name="search"
      placeholder="ex: Ho Chi Minh"
      type="search"
    />
    <button
      className="search__button"
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="content"
  >
    <div
      className="content__header"
    >
      <h1
        className="content__header__title"
      >
        Ho Chi Minh
      </h1>
      <p
        className="content__header__date"
      >
        <strong>
          Date: 
        </strong>
        01/01/2021
      </p>
    </div>
    <div
      className="weather"
    >
      <h3
        className="weather__title"
      >
        Today
      </h3>
      <img
        alt="info.label"
        className="weather__img"
      />
      <p>
        <strong>
          Max: 
        </strong>
        NaN
        °C
      </p>
      <p>
        <strong>
          Min: 
        </strong>
        NaN
        °C
      </p>
    </div>
    <div
      className="content__body"
    />
  </div>
</div>
`;
